spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java

```diff
@@ -5,27 +5,50 @@
  */
 package com.microsoft.azure.spring.cloud.config;
 
-import com.microsoft.azure.spring.cloud.autoconfigure.telemetry.TelemetryAutoConfiguration;
 import com.microsoft.azure.spring.cloud.autoconfigure.telemetry.TelemetryCollector;
+import com.microsoft.azure.spring.cloud.config.msi.AzureConfigMSIConnector;
+import com.microsoft.azure.spring.cloud.config.msi.ConfigAccessKeyResource;
+import com.microsoft.azure.spring.cloud.config.msi.ConfigMSICredentials;
 import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.impl.client.HttpClients;
-import org.springframework.boot.autoconfigure.AutoConfigureBefore;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
 
 import javax.annotation.PostConstruct;
 
 @Configuration
-@AutoConfigureBefore(TelemetryAutoConfiguration.class)
 @EnableConfigurationProperties(AzureCloudConfigProperties.class)
 @ConditionalOnClass(AzureConfigPropertySourceLocator.class)
 @ConditionalOnProperty(prefix = AzureCloudConfigProperties.CONFIG_PREFIX, name = "enabled", matchIfMissing = true)
 public class AzureConfigBootstrapConfiguration {
     private static final String AZURE_CONFIG_STORE = "AzureConfigService";
 
+    @Bean
+    @Primary
+    public AzureCloudConfigProperties initProperties(AzureCloudConfigProperties properties) {
+        if (StringUtils.hasText(properties.getConnectionString())) {
+            return properties;
+        }
+
+        ConfigMSICredentials msiCredentials = new ConfigMSICredentials(properties.getMsi());
+        ConfigAccessKeyResource keyResource = new ConfigAccessKeyResource(properties.getMsi());
+
+        AzureConfigMSIConnector msiConnector = new AzureConfigMSIConnector(msiCredentials, keyResource);
+        properties.setConnectionString(msiConnector.getConnection());
+        properties.validateAndInit();
+
+        Assert.hasText(properties.getConnectionString(), "Connection string cannot be empty");
```

Comment:
I think you should move this before setConnectionString.

Reply:
Will keep it as-is for now, to avoid an extra variable for `msiConnector.getConnection()`.